concordium-sdk/src/main/java/com/concordium/sdk/ClientV2.java

```diff
@@ -881,6 +887,79 @@ public ImmutableList<WinningBaker> getWinningBakersEpoch(EpochQuery epochQuery)
+        }
+        return listenForFinalizedTransaction(transactionHash, timeoutMillis);
+    }
+
+    /**
+     * Helper function for {@link ClientV2#waitUntilFinalized(Hash, int)}. Listens for finalized blocks and returns the {@link FinalizedBlockItem} of the transaction when available.
+     * @param transactionHash the {@link Hash} of the transaction to wait for.
+     * @param timeoutMillis the number of milliseconds to wait before throwing an exception.
+     * @return {@link FinalizedBlockItem} of the transaction.
+     * @throws io.grpc.StatusRuntimeException with {@link io.grpc.Status.Code}: <ul>
+     * <li> {@link io.grpc.Status#NOT_FOUND} if the transaction is not known to the node after the timeout is exceeded.
+     * <li> {@link io.grpc.Status#DEADLINE_EXCEEDED} if the transaction is not finalized after the timeout is exceeded.
+     * </ul>
+     */
+    private FinalizedBlockItem listenForFinalizedTransaction(Hash transactionHash, int timeoutMillis) {
+        io.grpc.StatusRuntimeException notFound = null;
+        try {
+            Iterator<BlockIdentifier> finalizedBlockStream = this.getFinalizedBlocks(timeoutMillis);
+            while (finalizedBlockStream.hasNext()) {
+                finalizedBlockStream.next();
+                // We check if the transaction is finalized every time a new block is finalized.
+                try {
+                    BlockItemStatus newStatus = this.getBlockItemStatus(transactionHash);
+                    if (newStatus.getStatus() == Status.FINALIZED) {
+                        return newStatus.getFinalizedBlockItem().get();
+                    }
+                } catch (io.grpc.StatusRuntimeException e) {
+                    ignoreExceptionIfCode(e, io.grpc.Status.Code.NOT_FOUND);
+                    notFound = e; // If the transaction is still not found after the timeout is exceeded, we throw this exception.
+                }
+            }
+        } catch (io.grpc.StatusRuntimeException e) {
+            ignoreExceptionIfCode(e, io.grpc.Status.Code.DEADLINE_EXCEEDED);
+            if (notFound != null ) { // The block was not found.
+                throw notFound;
+            }
+            throw e;
+        }
+        throw new IllegalStateException("No more finalized blocks"); // This should never happen as finalized blocks keep being streamed.
+    }
+
+    /**
+     * Helper method for allowing {@link io.grpc.StatusRuntimeException} with specified {@link io.grpc.Status.Code} without throwing the exception.
+     * @param e the {@link io.grpc.StatusRuntimeException}.
+     * @param code the {@link io.grpc.Status.Code} to allow.
+     * @throws io.grpc.StatusRuntimeException if the code of the exception does not match the provided code.
+     */
+    private void ignoreExceptionIfCode(io.grpc.StatusRuntimeException e, io.grpc.Status.Code code) {
+        if (e.getStatus().getCode().equals(code)) {
+            return;
+        }
+        throw e;
+    }
+
```

Comment:
```suggestion
     public Optional<FinalizedBlockItem> waitUntilFinalized(Hash transactionHash, int timeoutMillis) {
        Optional<FinalizedBlockItem> maybeStatus = getFinalizedTransaction(transactionHash);
        // if it's finalized return it.
        if (maybeStatus.isPresent()) {
            return maybeStatus;
        }
        // check newly finalized blocks until we time out
        Optional<FinalizedBlockItem> result = Optional.empty();
        try {
            Iterator<BlockIdentifier> finalizedBlockStream = this.getFinalizedBlocks(timeoutMillis);
            while (finalizedBlockStream.hasNext()) {
                finalizedBlockStream.next();
                Optional<FinalizedBlockItem> finalizedBlockItem = getFinalizedTransaction(transactionHash);
                if (finalizedBlockItem.isPresent()) {
                    // the transaction is included in a finalized block, break and return the finalized status.
                    result = finalizedBlockItem;
                    break;
                }
            }
        } catch (io.grpc.StatusRuntimeException e) {
            // we timed out. Return empty to indicate that the transaction could not be found.
            if (e.getStatus().getCode().equals(Status.Code.DEADLINE_EXCEEDED)) {
                return Optional.empty();
            }
            throw e;
        }
        return result;
    }

    private Optional<FinalizedBlockItem> getFinalizedTransaction(Hash transactionHash) {
        try {
            BlockItemStatus status = this.getBlockItemStatus(transactionHash);
            if (status.getFinalizedBlockItem().isPresent()) {
                return Optional.of(status.getFinalizedBlockItem().get());
            }
            // Only return a finalized transaction.
            return Optional.empty();
        } catch (io.grpc.StatusRuntimeException e) {
            // If the transaction is not found then return empty.
            if (e.getStatus().getCode().equals(io.grpc.Status.Code.NOT_FOUND)) {
                return Optional.empty();
            }
            // report back any other exceptions.
            throw e;
        }
    }

```

I think that this can be written a bit clearer and also I think we should return "empty" in the case where the transaction was not found.